### PR TITLE
Masked backtest

### LINKED ISSUE #563: Add masked backtest
### 🚀 Feature Request

Add ability to set masks for folds in backtest 

### Motivation

More flexibility for backtest strategy 

### Proposal

1. Add class `FoldMask`(see etna-utils)
2. In `AbstractPipeline`
- change the signature of `backtest`: `n_folds: Union[int, List[FoldMask]] = 5`
3. In `BasePipeline`
- change the signature of `backtest`: `n_folds: Union[int, List[FoldMask]] = 5`
- use the backtest implementation from etna-utils
- if `n_folds` is int, create fold masks for the `n_folds`
4. In all the places where backtest is used make `n_folds: Union[int, List[FoldMask]]`

### Test cases

Add tests from etna-utils

### Alternatives

_No response_

### Additional context

Blocked by #562

### Checklist

- [ ] I discussed this issue with ETNA Team

### PR DESCRIPTION
**Pull request by alex-hse-repository**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [ ] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Closing issues

closes #563

### CODE REVIEW

File: etna/pipeline/base.py [hunk 1]
```diff
@@ -24,6 +25,87 @@
...
+
+class CrossValidationMode(Enum):
+    """Enum for different cross-validation modes."""
+
+    expand = "expand"
+    constant = "constant"
+
+
+class FoldMask(BaseMixin):
+    """FoldMask class for masks."""
```
Comment (Mr-Geekman) [inline etna/pipeline/base.py:39]: Please, add a more precise description of this class. It is really not obvious what it does by this small line.

File: etna/pipeline/base.py [hunk 2]
```diff
@@ -24,6 +25,87 @@
...
+
+    def _validate_last_train_timestamp(self):
+        """Check that last train timestamp is later then first train timestamp."""
+        if self.first_train_timestamp and self.last_train_timestamp < self.first_train_timestamp:
+            raise ValueError("Last train timestamp should be strictly later than first train timestamp!")
+
+    def _validate_target_timestamps(self):
+        """Check that all target timestamps are later then last train timestamp."""
+        first_target_timestamp = self.target_timestamps[0]
+        if first_target_timestamp < self.last_train_timestamp:
```
Comment (Mr-Geekman) [inline etna/pipeline/base.py:73]: What about ==? Isn't <= more suitable here?

File: etna/pipeline/base.py [hunk 3]
```diff
@@ -24,6 +25,87 @@
...
+        dataset_timestamps = list(ts.index)
+        dataset_description = ts.describe()
+
+        min_first_timestamp = dataset_description["start_timestamp"].min()
+        if self.first_train_timestamp and self.first_train_timestamp < min_first_timestamp:
+            raise ValueError(f"First train timestamp should be later than {min_first_timestamp}!")
+
+        last_timestamp = dataset_description["end_timestamp"].min()
+        if self.last_train_timestamp > last_timestamp:
+            raise ValueError(f"Last train timestamp should be sooner than {last_timestamp}!")
```
Comment (Mr-Geekman) [inline etna/pipeline/base.py:95]: Not precise error message. The error occurs only if last_train_timestamp is later that last_timestamp. It means that last_train_timestamp should be not later then last_timestamp, not sooner.

File: etna/pipeline/base.py [hunk 4]
```diff
@@ -24,6 +25,87 @@
...
+
+        dataset_first_target_timestamp = dataset_timestamps[dataset_timestamps.index(self.last_train_timestamp) + 1]
+        mask_first_target_timestamp = self.target_timestamps[0]
+        if mask_first_target_timestamp < dataset_first_target_timestamp:
+            raise ValueError(f"First target timestamp should be later than {dataset_first_target_timestamp}!")
```
Comment (Mr-Geekman) [inline etna/pipeline/base.py:100]: Check borders

File: tests/test_pipeline/test_pipeline.py [hunk 5]
```diff
@@ -356,3 +362,130 @@ def test_forecast_pipeline_with_nan_at_the_end(df_with_nans_in_tails):
...
+            ],
+        ),
+    ),
+)
+def test_generate_masks_from_n_folds(example_tsds: TSDataset, n_folds, mode, expected_masks):
+    masks = Pipeline._generate_masks_from_n_folds(ts=example_tsds, n_folds=n_folds, horizon=3, mode=mode)
+    for mask, expected_mask in zip(masks, expected_masks):
+        assert mask.first_train_timestamp == expected_mask.first_train_timestamp
+        assert mask.last_train_timestamp == expected_mask.last_train_timestamp
+        assert [timestamp for timestamp in mask.target_timestamps] == [
```
Comment (Mr-Geekman) [inline tests/test_pipeline/test_pipeline.py:409]: Why do we need iteration here? Why can't we compare lists?

File: tests/test_pipeline/test_pipeline.py [hunk 6]
```diff
@@ -356,3 +362,130 @@ def test_forecast_pipeline_with_nan_at_the_end(df_with_nans_in_tails):
...
+
+
+@pytest.mark.parametrize(
+    "mask,expected",
+    (
+        (FoldMask("2020-01-01", "2020-01-07", ["2020-01-10"]), {"segment_0": 0, "segment_1": 11}),
+        (FoldMask("2020-01-01", "2020-01-07", ["2020-01-08", "2020-01-11"]), {"segment_0": 95.5, "segment_1": 5}),
+    ),
+)
+def test_run_fold(ts_run_fold: TSDataset, mask: FoldMask, expected: Dict[str, List[float]]):
```
Comment (Mr-Geekman) [inline tests/test_pipeline/test_pipeline.py:483]: Shouldn't it be before testing backtest method?

File: tests/test_pipeline/test_pipeline.py [hunk 7]
```diff
@@ -356,3 +362,130 @@ def test_forecast_pipeline_with_nan_at_the_end(df_with_nans_in_tails):
...
+    metrics = dict(metrics_df.values)
+    for segment in expected.keys():
+        assert segment in metrics.keys()
+        np.testing.assert_array_almost_equal(expected[segment], metrics[segment])
+
+
+@pytest.mark.parametrize(
+    "lag,expected", ((4, {"segment_0": 0, "segment_1": 0}), (7, {"segment_0": 0, "segment_1": 0.5}))
+)
+def test_backtest_right_timestamps(masked_ts: TSDataset, lag: int, expected: Dict[str, List[float]]):
```
Comment (Mr-Geekman) [inline tests/test_pipeline/test_pipeline.py:462]: What is the difference with previous test?

File: etna/pipeline/base.py (hunk 1 above)
Comment (alex-hse-repository) [reply etna/pipeline/base.py:39]: Fixed, got the this description from etna-utils)

File: tests/test_pipeline/test_pipeline.py (hunk 5 above)
Comment (alex-hse-repository) [reply tests/test_pipeline/test_pipeline.py:409]: Fixed, in the first version I casted the elements to timestamps)

File: tests/test_pipeline/test_pipeline.py (hunk 6 above)
Comment (alex-hse-repository) [reply tests/test_pipeline/test_pipeline.py:483]: Fixed, got the order from etna-utils)

File: tests/test_pipeline/test_pipeline.py (hunk 7 above)
Comment (alex-hse-repository) [reply tests/test_pipeline/test_pipeline.py:462]: I have no guess) The only difference in the parameters here

File: etna/pipeline/base.py [hunk 8]
```diff
@@ -24,6 +25,89 @@
...
+class CrossValidationMode(Enum):
+    """Enum for different cross-validation modes."""
+
+    expand = "expand"
+    constant = "constant"
+
+
+class FoldMask(BaseMixin):
+    """Container to hold the description of the fold mask.
+    Fold masks are expected to be used for backtest strategy customization.
```
Comment (Mr-Geekman) [inline etna/pipeline/base.py:40]: May be we should add empty line here according to https://numpydoc.readthedocs.io/en/latest/format.html
Example that I've found: https://github.com/numpy/numpy/blob/v1.22.0/numpy/lib/nanfunctions.py#L953-L1056

File: tests/test_pipeline/test_pipeline.py (hunk 7 above)
Comment (Mr-Geekman) [reply tests/test_pipeline/test_pipeline.py:478]: May be we should change the name of this task somehow.
For example, previous will be `test_backtest_one_point`, the second `test_backtest_two_points`.
Or we should just ask the creator of this tests.

**Review by iKintosh — APPROVED** [review]

File: tests/test_pipeline/test_pipeline.py (hunk 7 above)
Comment (alex-hse-repository) [reply tests/test_pipeline/test_pipeline.py:478]: I renamed the tests

**Review by iKintosh — APPROVED** [review]

**Pull request merged**